docs/proposals/20221030-garbage-collection-of-tenant-resources.md

```diff
@@ -0,0 +1,153 @@
+
+### Non-Goals/Future Work
+
+- When a tenant creates a resource through KubeZoo, add a label to it. Therefore, we can use label selectors to filter the resources of a tenant to improve performance when performing garbage collection. This optimization will be designed and implemented in the follow-up.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+Tenants can create some native cluster-scoped resources through KubeZoo. For example,
+
+- tenants can manage storage by manually creating PersistentVolumes;
+
+- tenants can define some specific PriorityClasses to represent the priority of a Pod when queuing and preempting;
+
+- Tenants can build custom admission webhooks (not yet supported) and configure them through MutatingWebhookConfigurations or ValidatingWebhookConfigurations.
+
+When a tenant is deleted, the tenant's native cluster-scoped resources should be cleaned up.
+
+#### Story 2
+
+Tenants can create some CRDs through KubeZoo to extend their API definitions. When a tenant is deleted, the CRDs defined by the tenant should also be cleaned up.
+
+#### Story 3
+
+Tenants can create some cluster-scoped CRs through KubeZoo, which may be of the following two types:
+
+- tenant-defined CRDs
+- system CRDs predefined by the admin
+
+When a tenant is deleted, CRs of the above kinds should also be cleaned up.
+
+## API
+
+Introduce a finalizer for the Tenant object:
+
+```yaml
+apiVersion: tenant.kubezoo.io/v1alpha1
+kind: Tenant
+metadata:
+  name: "111111"
+  finalizers:
+  - kubezoo.io/tenant
+spec:
+  id: 111111
+```
+
+### Design Details
+
+Modify the Tenant Controller's processing logic for Tenant events:
+
+- When a `CREATE` event is watched, add a `kubezoo.io/tenant` finalizer for the Tenant object.
+
+- When an `UPDATE` event is watched, if the Tenant has a `DeletionTimestamp` and a `kubezoo.io/tenant` finalizer,
+  
+    1. List all cluster-scoped GVRs via DiscoveryClient, and divide them into two groups according to whether they are CRD or not.
+       
+    2. Traverse the GVR list of CRDs. For each GVR, check whether the CRD belongs to this tenant according to the prefix of its `Group`. If so, delete it via DynamicClient.
+
+        > In this step, CRs associated with CRDs belonging to this tenant will be deleted in cascade. This avoids the overhead of listing all CRs in step (iii) and improves performance.
+    
+    3. Traverse the list of non-CRD GVRs. For each GVR, list all objects of this kind via DynamicClient. For each object, check whether it belongs to this tenant according to the prefix of its `Name`. If so, delete it via DynamicClient.
+        > - In this step, the tenant's namespaces will be cleaned up, which triggers the Namespace Controller to clean up the tenant's Namespaced resources.
+        > - In this step, the tenant's CRs associated with system CRDs are also cleaned up.
+
+    4. Remove the `kubezoo.io/tenant` finalizer from the Tenant object.
```

Comment:
Is it necessary to wait for all associated objects to be actually deleted first? Otherwise there may be race condition when a tenant with the same name is immediately recreated.

Reply:
Yes, I think this race condition does exist. For each GVR, we can check if there are any remaining objects of that tenant after the deletion in steps (ii) and (iii).